frontend/src/toolbar/elements/heatmapLogic.ts

```diff
@@ -152,9 +160,16 @@ export const heatmapLogic = kea<heatmapLogicType>([
                         combinedSelector = lastSelector ? `${selector} > ${lastSelector}` : selector
 
                         try {
-                            const domElements = Array.from(
-                                querySelectorAllDeep(combinedSelector, document, allElements)
-                            ) as HTMLElement[]
+                            if (cache.selectorToElement === undefined) {
+                                cache.selectorToElement = {}
+                            }
+                            let domElements: HTMLElement[] = cache.selectorToElement?.[combinedSelector] || []
+                            if (domElements.length === 0) {
+                                domElements = Array.from(
+                                    querySelectorAllDeep(combinedSelector, document, pageElements)
+                                ) as HTMLElement[]
+                                cache.selectorToElement[combinedSelector] = domElements
+                            }
```

Comment:
Are you explicitly searching again all elements that had 0 matches? If so, worth making this point clear with a comment. Otherwise:

```suggestion
                            let domElements: HTMLElement[] = cache.selectorToElement?.[combinedSelector] || []
                            if (typeof domElements === 'undefined') {
                                domElements = Array.from(
                                    querySelectorAllDeep(combinedSelector, document, pageElements)
                                ) as HTMLElement[]
                                cache.selectorToElement[combinedSelector] = domElements
                            }
```

Reply:
🧠